duetector/tracers/tcpconnect.py

```diff
@@ -0,0 +1,135 @@
+from collections import namedtuple
+from typing import Callable, NamedTuple
+
+from duetector.extension.tracer import hookimpl
+from duetector.tracers.base import BccTracer
+
+
+class TcpconnectTracer(BccTracer):
+    """
+    A tracer for openat2 syscall
+    """
+
+    default_config = {
+        **BccTracer.default_config,
+        "poll_timeout": 10,
+    }
+
+    many_attatchs = [
+        ("kprobe", {"fn_name": "do_trace", "event": "tcp_v4_connect"}),
+        ("kretprobe", {"fn_name": "do_return", "event": "tcp_v4_connect"}),
+    ]
+
+    poll_fn = "ring_buffer_poll"
+
+    @property
+    def poll_args(self):
+        return {"timeout": int(self.config.poll_timeout)}
+
+    data_t = namedtuple("TcpTracking", ["pid", "comm", "saddr", "daddr", "dport"])
+
+    # define BPF program
+    prog = """
+    #include <uapi/linux/ptrace.h>
+    #include <net/sock.h>
+    #include <bcc/proto.h>
+    #define TASK_COMM_LEN 16
+
+    BPF_RINGBUF_OUTPUT(buffer, 1 << 4);
+    BPF_HASH(currsock, u32, struct sock *);
+
+    struct event {
+        u32 dport;
+        u32 saddr;
+        u32 daddr;
+        u32 pid;
+        char comm[TASK_COMM_LEN];
+    };
+    int do_trace(struct pt_regs *ctx, struct sock *sk)
+    {
+	    u32 pid = bpf_get_current_pid_tgid();
+
+	    // stash the sock ptr for lookup on return
+	    currsock.update(&pid, &sk);
+
+	    return 0;
+    }
+
+    int do_return(struct pt_regs *ctx)
+    {
+	    int ret = PT_REGS_RC(ctx);
+	    u32 pid = bpf_get_current_pid_tgid();
+        struct event event= {};
+
+	    struct sock **skpp;
+	    skpp = currsock.lookup(&pid);
+	    if (skpp == 0) {
+		    return 0;	// missed entry
+	    }
+
+	    if (ret != 0) {
+		    // failed to send SYNC packet, may not have populated
+		    // socket __sk_common.{skc_rcv_saddr, ...}
+		    currsock.delete(&pid);
+		    return 0;
+	    }
+
+	    // pull in details
+	    struct sock *skp = *skpp;
+	    u32 saddr = skp->__sk_common.skc_rcv_saddr;
+	    u32 daddr = skp->__sk_common.skc_daddr;
+	    u16 dport = skp->__sk_common.skc_dport;
+        event.saddr = saddr;
+        event.daddr = daddr;
+        event.dport = dport;
+        event.pid = pid;
+        bpf_get_current_comm(&event.comm, sizeof(event.comm));
+	    // output
+	    buffer.ringbuf_output(&event, sizeof(event), 0);
+	    //bpf_trace_printk("trace_tcp4connect %x %x %d\\n", saddr, daddr, ntohs(dport));
+
+	    currsock.delete(&pid);
+
+	    return 0;
+    }
+    """
+
+    def set_callback(self, host, callback: Callable[[NamedTuple], None]):
+        def _(ctx, data, size):
+            event = host["buffer"].event(data)
+            return callback(self._convert_data(event))  # type: ignore
+
+        host["buffer"].open_ring_buffer(_)
+
```

Comment:
I notice that the data here is actually unreadable, readable data needs to be processed by `inet_ntoa`, can we integrate `inet_ntoa` into `_convert_data` ?

Reply:
I have integrated `inet_ntoa` into `_convert_data` in my newest commit.

Reply:
You should overwrite `_convert_data` in subclass